### ISSUE TITLE
Add support of batch operations to URID extension

### ISSUE THREAD

**Issue by sadko4u** [drobilla/lv2#28]
Some hosts like Carla perform complete isolation of plugins: all plugins are launched as separate processes in the system so crash of one plugin will not affect other plugins. Additionally, all UIs are executed as separate processes, so the communication between UI and the plugin is performed via standard system IPC that is slow.

That yields to the following case: when plugin requests many URIDs at instantiation time, all calls are synchronously executed via the IPC mechanism of the host, so each URID.map and URID.unmap call takes much longer time than expected. Launching multiple sequential requests performs **TIMES** slower than on hosts that shares the common address space with plugins.

That's why **very reasonable** would be implementation of extensions for the URID interface that allow batch operations with URIDs. The following gist shows an prototype of such interface, as a part of the original **<lv2/ns/ext/urid/urid.h>** header:

https://gist.github.com/sadko4u/8f11ae5632424c30c0578ba233a899a4

Here's the related issue:

https://github.com/sadko4u/lsp-plugins/issues/25

Also It would be nice to see any suggestions related to this feature from @x42 and @falkTX .

**Comment by x42** [comment]
My opinion is that process isolating plugins makes no sense. It only adds unreasonable overhead.

If a plugin that processes audio crashes, there will be audible artifacts. Game Over. It now no longer matters if other plugins are affected or not.

**Comment by falkTX** [comment]
the DSP is not bridged, just the UI, as that is the biggest reason of crashes.

**Comment by x42** [comment]
> the DSP is not bridged, just the UI, as that is the biggest reason of crashes.
can the DSP survive in a sane state if the UI crashes?

**Comment by falkTX** [comment]
yes. of course. it is as if the UI got closed by the user.
DSP side knows the full state, when user opens UI again, the UI receives the full parameter/state again

**Comment by x42** [comment]
@sadko4u  did you investigate speeding up the IPC mechanism itself? If that is indeed a bottleneck, I'd expect more issues, not simpy URID mapping.. That is usually a one-time operation.

I also recall @drobilla considering to hardcode some common URIDs (in particular those that the forge uses regularly)

**Comment by falkTX** [comment]
IPC could be optimized a lot more, pipes are slow by design compared to other things...
(I had the idea of porting the fast shared-memory sync code used for audio to these ui bridges)

but anyway 500+ URID maps on init is a bit much.
being able to batch these operations would be nice regardless of carla, bridges and whatever else.

**Comment by drobilla** [comment]
I think a host that wants to do this should implement an appropriate URI map that accounts for this, by caching, or mapping most up front, for example.  This would dramatically reduce the amount of IPC needed in most cases, often to zero.  I've considered doing this for ingen but never gotten around to it for lack of need.

It seems inappropriate to complicate the extension API to support a poor implementation of an esoteric use case, especially when it wouldn't help with existing plugins anyway, and require a bunch of porting for just this reason.

I'd be more open to this if it was a part of work that made things nice in general (maybe standard strict in extensions that are all mapped at once, sort of like the forge is abused for now, or the static thing) that also moved towards this benefit.

**Comment by sadko4u** [comment]
As @falkTX mentioned, the plugin UI requires a lot URIDs to be mapped at instantiation time. It can not be optimized because is already done in a cycle. The bottleneck is the IPC between host and UI. Batching such operations is a good way to optimize the interaction between UI and host.

As you said, 10 URD mappings won't give any sense. But 500+ URDs sure will. Even more, because URIDs are the main identification mechanism that is used anywhere, I think it should be a good mechanism to retrieve them fast, even with slow communication channel.

**Comment by x42** [comment]
Are you sure that batch-processing does make a difference for hosts that don't use IPC?

For hosts that do not use IPC, the proposed batch-processing may even be  tad slower since you need to allocate and free memory for the array to be batch processed, and the plugin needs to provide a free callback (for windows to share heap between dlls).  While I expect context-free function call overhead to be negligible.

Also most plugins that I've seen so far don't dynamically map that many URIDs, so it's unlikely any other plugins would check if a host supports this and optionally use the batch-API.

Regarding Carla.  Since IPC is used for general DSP <> UI communication there, not only URID mapping, perhaps that should be optimized first.  Realistically, a single request on localhost should take less than a msec (compare to what HTTP servers do).  @falkTX  do you think ~1ms per request is realistic for Carla, too?

**Comment by sadko4u** [comment]
>For hosts that do not use IPC, the proposed batch-processing may even be tad slower since you need to allocate and free memory for the array to be batch processed
If you look at the code snippet I've provided, the host is not required to allocate the memory. The plugin provides array of strings and array to store URIDs, so actually host needs just to store valid URIDs to this array.

>and the plugin needs to provide a free callback (for windows to share heap between dlls)
No, it doesn't. Or I've missed something?

**Comment by drd-ableton** [comment]
Yesh, I understand the problem.  It can be optimized, because you can, for example, copy the existing map to the other process _first_ and only do fine-grained IPC for new stuff.  In combination with a sensibly pre-populated map, this would all but eliminate this flurry of round trips.

Again, I think there is a bunch of stuff that would improve things in general and perhaps move in this direction which would be nice to see.  This, however, just _adds_ complexity to plugins, in an area where we need to do exactly the opposite.  It _requires_ any plugin and/or UI to be rewritten to use this strange interface, with the only benefit being that it instantiates more quickly in this weird scenario.  The reality is that every single existing UI that maps URIs at all does it with the existing function, and will for a long time, if not ever.

Plus, plugins that are properly designed to work in remote situations don't require the same URI map in UI and instance at all.  There's a reason atoms are host transparent and can be serialized.  The only reason it would _have_ to be that way is if they have dirty raw pointers into each other (via instance-access, for example, or other nasty behind the scenes tricks), in which case the goal of crash prevention is pointless anyway.

In summary, sorry, but this causes a lot of problems in the spec that must be used by everyone, to work around something in a host that seems massively slower than it could be (even with the given design), in a design that is a bit questionable and contrary to many of the core design decisions of LV2.  I'd suggest improving your IPC performance first, since that will actually improve the situation right now regardless, then considering other optimizations, before this.

If it does get to that point, there should be a solid answer to the question "why are you doing round-trips to map URIDs that were already mapped when the process/UI was launched anyway?"  Surely if you can implement a batch URI mapping protocol, you can also implement sending over the existing map first.

**Comment by drobilla** [comment]
Yes, I understand the problem.  It can be optimized, because you can, for example, copy the existing map to the other process _first_ and only do fine-grained IPC for new stuff.  In combination with a sensibly pre-populated map, this would all but eliminate this flurry of round trips.

Again, I think there is a bunch of stuff that would improve things in general and perhaps move in this direction which would be nice to see.  This, however, just _adds_ complexity to plugins, in an area where we need to do exactly the opposite.  It _requires_ any plugin and/or UI to be rewritten to use this strange interface, with the only benefit being that it instantiates more quickly in this weird scenario.  The reality is that every single existing UI that maps URIs at all does it with the existing function, and will for a long time, if not ever.

Plus, plugins that are properly designed to work in remote situations don't require the same URI map in UI and instance at all.  There's a reason atoms are host transparent and can be serialized.  The only reason it would _have_ to be that way is if they have dirty raw pointers into each other (via instance-access, for example, or other nasty behind the scenes tricks), in which case the goal of crash prevention is pointless anyway.

In summary, sorry, but this causes a lot of problems in the spec that must be used by everyone, to work around something in a host that seems massively slower than it could be (even with the given design), in a design that is a bit questionable and contrary to many of the core design decisions of LV2.  I'd suggest improving your IPC performance first, since that will actually improve the situation right now regardless, then considering other optimizations, before this.

If it does get to that point, there should be a solid answer to the question "why are you doing round-trips to map URIDs that were already mapped when the process/UI was launched anyway?"  Surely if you can implement a batch URI mapping protocol, you can also implement sending over the existing map first.

**Comment by sadko4u** [comment]
> It _requires_ any plugin and/or UI to be rewritten to use this strange interface, with the only benefit being that it instantiates more quickly in this weird scenario. The reality is that every single existing UI that maps URIs at all does it with the existing function, and will for a long time, if not ever.

It **doesn't require** UI or any plugin to be rewritten. The plugin/UI can utilize the previously well-defined URID:map and URID:unmap interface. Nobody asks to deprecate these interfaces. I'm asking to add **two new interfaces that can provide batch functionality** as an addition to URID:map and URID:unmap. Most hosts even don't need to implement these interfaces. But hosts that have bottleneck with IPC can do this as a fast solution for the serious problem.

> Plus, plugins that are properly designed to work in remote situations don't require the same URI map in UI and instance at all. There's a reason atoms are host transparent and can be serialized.

It also **depends** on the host that launches UI and DSP code.

> In summary, sorry, but this causes a lot of problems in the spec that must be used by everyone

No, as I previously described, it **should not** be used by everyone. Moreover, plugins that do not use extensively URID:map won't meet any problems. The main problem is related to plugins that use URID:map extensively at the **instantiation stage** which yields to long instantiation delay which is annoying for several plugin users.

>I'd suggest improving your IPC performance first, since that will actually improve the situation right now regardless, then considering other optimizations, before this.

The IPC of LV2 is always bottleneck because Atoms ports are slow and very complicated to use. Even for Ardour transferring of ~100k byte data @ 25 Hz rate is a really hard task, so the plugin developers are forced to use 'dirty' and 'tricky' LV2:Instance interface if it is possible and the host provides such capability.

> If it does get to that point, there should be a solid answer to the question "why are you doing round-trips to map URIDs that were already mapped when the process/UI was launched anyway?"

Is it a question to host developer or plugin developer? As a plugin developer I just use URID:map at the instantiation time, but it works damn slowly. And then, I get a bug report that my UI is slowly working under certain hosts. And this is a result of that fact that I am required to use Atoms to transfer plugin state between UI and the plugin to conform the LV2 standard. And the transport is slow. And finally all gets worse.


 Surely if you can implement a batch URI mapping protocol, you can also implement sending over the existing map first.

**Comment by falkTX** [comment]
A note here from Carla. Carla already does what you suggest, it sends the mapped URIDs to the UI process when it starts. In https://github.com/falkTX/Carla/blob/develop/source/backend/plugin/CarlaPluginLV2.cpp#L1443
Only new urid map calls have to be transferred over.

There is an issue of the UI wanting to map a lot of URIDs that were not needed beforehand.
I guess initializing those on the DSP side would help things a bit, but it is also a bit wasteful for cases where the UI is never opened.

**Comment by sadko4u** [comment]
Moreover, in LSP, all URIDS are taken in DSP code once at instantiation time:
https://github.com/sadko4u/lsp-plugins/blob/1d41a52f6b2236f38670c2d3a28ab97c419a4754/include/container/lv2/ports.h#L25
and:
https://github.com/sadko4u/lsp-plugins/blob/1d41a52f6b2236f38670c2d3a28ab97c419a4754/include/container/lv2/extensions.h#L465
https://github.com/sadko4u/lsp-plugins/blob/1d41a52f6b2236f38670c2d3a28ab97c419a4754/include/container/lv2/extensions.h#L448

**Comment by ventosus** [comment]
@sadko4u, I've given your plugin UIs a spin on my crappy synthpod host [1]. It runs each UI in a separate process, too. synthpod has a separate URID hash per process, though (like @drobilla explained above) and serializes atoms via shared memory. This is real quick.

Your LV2 plugin UI's (after adding a bloody hack to fix #32), open up instantaneously on synthpod.

So, your issue is not specific for hosts that run UIs in separate processes.
You seem to do everyting correctly.
There's really nothing to fix in the LV2 URID spec, imho.
The only thing that needs fixing is one-or-another *implementation* of the LV2 URID spec :)

[1] https://git.open-music-kontrollers.ch/lv2/synthpod

**Comment by sadko4u** [comment]
@ventosus thank you for information. I'll fix the #32 issue ASAP. It's curious how I've missed this and kept for a long time being not changed.
So it seems that implementation of LV2 URID is a bottleneck in Carla.

**Comment by drobilla** [comment]
> No, as I previously described, it **should not** be used by everyone. Moreover, plugins that do not use extensively URID:map won't meet any problems. The main problem is related to plugins that use URID:map extensively at the **instantiation stage** which yields to long instantiation delay which is annoying for several plugin users.

This is the usual use case of URID mapping.  Pretty much the only relevant one here, really: doing this is the whole point of the API.
